Manipulate csr_cycle and PC by registers. Drop insn_len

qwe661234 opened the pull request:
To eliminate the overhead of memory operation on frequently used rv
fields and ir fields, we pass csr_cycle and PC as parameter and assigin
insn_len as constant value.

* Before
```
000000000000b1f0 <do_nop>:
    b1f0: addq $0x1,0x1a8(%rdi)
    b1f8: movl $0x0,0x58(%rdi)
    b1ff: movzbl 0x1c(%rsi),%eax
    b203: add %eax,0xd8(%rdi)
    b209: movzbl 0x1d(%rsi),%eax
    b20d: or 0x110(%rdi),%al
    b213: jne b220 <do_nop+0x30>
    b215: mov 0x38(%rsi),%rsi
    b219: jmpq *0x20(%rsi)
    b21c: nopl 0x0(%rax)
    b220: retq
    b221: nopw %cs:0x0(%rax,%rax,1)
    b22c: nopl 0x0(%rax)
```
* After
```
000000000000a6d0 <do_nop>:
    a6d0: movl $0x0,0x58(%rdi)
    a6d7: movzbl 0x1d(%rsi),%eax
    a6db: add $0x1,%rdx
    a6df: add $0x4,%ecx
    a6e2: or 0x110(%rdi),%al
    a6e8: jne a6f8 <do_nop+0x28>
    a6ea: mov 0x38(%rsi),%rsi
    a6ee: jmp *0x20(%rsi)
    a6f1: nopl 0x0(%rax)
    a6f8: mov %rdx,0x1a8(%rdi)
    a6ff: mov %ecx,0xd8(%rdi)
    a705: ret
    a706: cs nopw 0x0(%rax,%rax,1)
```
The operands of two x86 add instructions for csr_cycle, PC and insn_len
are all registers, instead of memory location.

Close: #241

jserv reviewed (dismissed):
Rebase the latest `master` branch where I made some stupid mistakes and reverted.

jserv commented on src/emulate.c:
```diff
@@ -368,17 +370,21 @@ static bool branch_taken = false;
-            return true;                                    \
-        const rv_insn_t *next = ir->next;                   \
-        MUST_TAIL return next->impl(rv, next);              \
+#define RVOP(inst, compressed, code)                                        \
```
I don't think it is a good idea to specify `compress` in parameter list. Can you avoid it?

qwe661234 replied:
Maybe we can hardcode PC += 4 and place PC -= 2 in all C-extension instruction.

jserv replied:
> Maybe we can hardcode PC += 4 and place PC -= 2 in all C-extension instruction.

No, the offset for the next PC should be determined within each instruction. Whether it's 4 or 2 bytes, this value should be predictable and consistent. Additionally, the support of RV32C is build-time configurable, which means that we should handle the PC in a general way.

qwe661234 replied:
Because we cannot get the information about the instruction is compressed or not, I only figure two ideas, the first is adding this information in parameter list, and the second is handling PC += 2 or 4 per function.

jserv replied:
> Because we cannot get the information about the instruction is compressed or not, I only figure two ideas, the first is adding this information in parameter list, and the second is handling PC += 2 or 4 per function.

How about extending `src/decode.h` by addressing compressed instructions?

qwe661234 replied:
However, the RVOP(...) marco is unrelated to `src/decode.h`.

jserv replied:
> However, the RVOP(...) marco is unrelated to `src/decode.h`.

We can utilize the X-macro trick to enumerate the length of each instruction. For example, `RV_INSN_LEN_addi = 4`. Then, the macro RVOP can directly employ concatenation in the C preprocessor to manage PC increment.

jserv commented on src/decode.h:
```diff
@@ -31,157 +31,157 @@ enum op_field {
-        _(cjalr, 1, ENC(rs1, rs2, rd))           \
-        _(cadd, 0, ENC(rs1, rs2, rd))            \
-        _(cswsp, 0, ENC(rs2))                    \
+#define RV_INSN_LIST                                \
```
Refine the above comment for correct format description.

jserv commented on src/emulate.c:
```diff
@@ -350,7 +352,15 @@ FORCE_INLINE bool insn_is_misaligned(uint32_t pc)
 enum {
-#define _(inst, can_branch, reg_mask) __rv_insn_##inst##_canbranch = can_branch,
+#define _(inst, can_branch, insn_len, reg_mask) \
+    __rv_insn_##inst##_insn_len = insn_len,
```
Simply rewrite as `__rv_insn_##inst##_len` to get rid of repeated `insn`.

jserv commented on src/emulate.c:
```diff
@@ -350,7 +352,15 @@ FORCE_INLINE bool insn_is_misaligned(uint32_t pc)
-#define _(inst, can_branch, reg_mask) __rv_insn_##inst##_canbranch = can_branch,
+#define _(inst, can_branch, insn_len, reg_mask) \
+    __rv_insn_##inst##_insn_len = insn_len,
+    RV_INSN_LIST
```
Drop `RV_INSN_LIST` here.

jserv commented:
Can you review all instances of `insn_len` (e.g., `src/decode.[ch]`) and determine if it still serves a meaningful purpose? It is possible that we can remove it to simplify the code.

jserv commented on src/emulate.c:
```diff
@@ -348,9 +350,11 @@ FORCE_INLINE bool insn_is_misaligned(uint32_t pc)
+#define _(inst, can_branch, insn_len, reg_mask) \
+    __rv_insn_##inst##_len = insn_len,          \
+    __rv_insn_##inst##_canbranch = can_branch,
     RV_INSN_LIST
```
This modification of single `enum` would provide `RV_INSN_LIST` an unexpected value. Instead, you must use two 'enum'. The term `RV_INSN_LIST` only appears once.

qwe661234 replied:
However, `enum` is outside the define,
```c
enum {
#define _(inst, can_branch, insn_len, reg_mask) \
    __rv_insn_##inst##_len = insn_len,
    RV_INSN_LIST
#undef _
};
```
Do you means a `enum` inside a `enum`?

jserv replied:
Ouch. I misinterpreted the code. No visible concern here.

jserv commented on src/emulate.c:
```diff
@@ -582,14 +628,14 @@ static void block_translate(riscv_t *rv, block_map_t *map, block_t *block)
         /* decode the instruction */
         if (!rv_decode(ir, insn)) {
-            rv->compressed = (ir->insn_len == INSN_16);
+            rv->compressed = ((insn & FC_OPCODE) != 3);
```
The enumeration of `INSN_16` and `INSN_32` can be removed from `src/decode.h`.

jserv replied:
Add an inline helper function in `src/decode.h` or `src/riscv_private.h` to perform `insn & FC_OPCODE) != 3`.

qwe661234 committed:
Unmap C.NOP to NOP

jserv commented on src/decode.h:
```diff
@@ -307,3 +300,5 @@ typedef struct rv_insn {
 /* decode the RISC-V instruction */
 bool rv_decode(rv_insn_t *ir, const uint32_t insn);
+
+#define CHECK_COMPRESSED_INSN ((insn & FC_OPCODE) != 3)
```
Rewrite as an inline function, so that compiler can check the types. In addition, write the corresponding comment for the detection of RV32C instructions.

jserv commented on src/decode.h:
```diff
@@ -307,3 +300,9 @@ typedef struct rv_insn {
 bool rv_decode(rv_insn_t *ir, const uint32_t insn);
+
+/* Detect the instruction is RV32C or not */
+FORCE_INLINE bool is_compressed(uint32_t insn)
```
This type of helper function is mostly concerned with implementation details and is best off not being placed in `src/decode.h` public header.

qwe661234 replied:
We can move it to src/riscv_private.h.

qwe661234 committed:
Remove insn_len field in IR structure

jserv reviewed (changes requested):
Rebase the latest `master` branch.

jserv commented on src/emulate.c:
```diff
@@ -348,9 +350,18 @@ FORCE_INLINE bool insn_is_misaligned(uint32_t pc)
     );
 }
 
+/* insn_len information for each RISC-V instruction */
```
Replace `insn_len` with instruction length in the comment.

qwe661234 committed:
Manipulate csr_cycle, PC and insn_len by register

To eliminate the overhead of memory operation on frequently used rv
fields and ir fields, we pass csr_cycle and PC as parameter and assigin
insn_len as constant value.

* Before

000000000000b1f0 <do_nop>:
    b1f0: 48 83 87 a8 01 00 00 addq $0x1,0x1a8(%rdi)
    b1f7: 01
    b1f8: c7 47 58 00 00 00 00 movl $0x0,0x58(%rdi)
    b1ff: 0f b6 46 1c movzbl 0x1c(%rsi),%eax
    b203: 01 87 d8 00 00 00 add %eax,0xd8(%rdi)
    b209: 0f b6 46 1d movzbl 0x1d(%rsi),%eax
    b20d: 0a 87 10 01 00 00 or 0x110(%rdi),%al
    b213: 75 0b jne b220 <do_nop+0x30>
    b215: 48 8b 76 38 mov 0x38(%rsi),%rsi
    b219: ff 66 20 jmpq *0x20(%rsi)
    b21c: 0f 1f 40 00 nopl 0x0(%rax)
    b220: c3 retq
    b221: 66 66 2e 0f 1f 84 00 data16 nopw %cs:0x0(%rax,%rax,1)
    b228: 00 00 00 00
    b22c: 0f 1f 40 00 nopl 0x0(%rax)

* After

000000000000a6d0 <do_nop>:
    a6d0: c7 47 58 00 00 00 00 movl $0x0,0x58(%rdi)
    a6d7: 0f b6 46 1d movzbl 0x1d(%rsi),%eax
    a6db: 48 83 c2 01 add $0x1,%rdx
    a6df: 83 c1 04 add $0x4,%ecx
    a6e2: 0a 87 10 01 00 00 or 0x110(%rdi),%al
    a6e8: 75 0e jne a6f8 <do_nop+0x28>
    a6ea: 48 8b 76 38 mov 0x38(%rsi),%rsi
    a6ee: ff 66 20 jmp *0x20(%rsi)
    a6f1: 0f 1f 80 00 00 00 00 nopl 0x0(%rax)
    a6f8: 48 89 97 a8 01 00 00 mov %rdx,0x1a8(%rdi)
    a6ff: 89 8f d8 00 00 00 mov %ecx,0xd8(%rdi)
    a705: c3 ret
    a706: 66 2e 0f 1f 84 00 00 cs nopw 0x0(%rax,%rax,1)
    a70d: 00 00 00

The operands of two x86 add instructions for csr_cycle, PC and insn_len
are all registers, instead of memory location.

Close: #241
